metronome/hotstuff/service/src/io/iohk/metronome/hotstuff/service/SyncService.scala

```diff
@@ -98,43 +137,63 @@ class SyncService[F[_]: Sync, N, A <: Agreement](
+          case Left(ex) =>
+            tracers.responseIgnored((from, response, Some(ex)))
+        }
     }
 
     process.handleErrorWith { case NonFatal(ex) =>
       tracers.error(ex)
     }
   }
 
-  /** Read Requests from the BlockSyncPipe and send Responses. */
-  def processBlockSyncPipe: F[Unit] = {
+  /** Read Requests from the BlockSyncPipe and send Responses.
+    *
+    * These are coming from the `ConsensusService` asking for a
+    * `Prepare` message to be synchronised with the sender.
+    */
+  private def processBlockSyncPipe(
+      blockSynchronizer: BlockSynchronizer[F, N, A]
+  ): F[Unit] = {
     blockSyncPipe.receive
-      .mapEval[Unit] { case request @ BlockSyncPipe.Request(sender, prepare) =>
-        // TODO (PM-3134): Block sync.
-        // TODO (PM-3132, PM-3133): Block validation.
-
-        // We must take care not to insert blocks into storage and risk losing
-        // the pointer to them in a restart. Maybe keep the unfinished tree
-        // in memory until we find a parent we do have in storage, then
-        // insert them in the opposite order, validating against the application side
-        // as we go along, finally responding to the requestor.
-        //
-        // It is enough to respond to the last block positively, it will indicate
-        // that the whole range can be executed later (at that point from storage).
-        val isValid: F[Boolean] = ???
-
-        isValid.flatMap { isValid =>
-          blockSyncPipe.send(BlockSyncPipe.Response(request, isValid))
-        }
+      .mapEval[Unit] {
+        // TODO (PM-3063): Change `BlockSyncPipe` to just `SyncPipe` and add
+        // ViewState sync requests which poll the fedreation for the latest
+        // Commit Q.C. and jump to it. When that signal comes, cancel the
+        // `syncFiberMap`, discard the `blockSynchronizer` and move over to
+        // state syncing, then create a new new block synchronizer and resume.
+        // For this, change the input of this method to a `F[BlockSynchronizer[F,N,A]]`
+        // and call some mutually recursive method representing different states:
+
+        case request @ BlockSyncPipe.Request(sender, prepare) =>
+          // It is enough to respond to the last block positively, it will indicate
+          // that the whole range can be executed later (at that point from storage).
+          // If the same leader is sending us newer proposals, we can ignore the
+          // previous pepared blocks - they are either part of the new Q.C.,
+          // in which case they don't need to be validated, or they have not
+          // gathered enough votes, and been superseded by a new proposal.
+          syncFiberMap.cancelQueue(sender) >>
+            syncFiberMap
+              .submit(sender) {
+                for {
+                  _       <- blockSynchronizer.sync(sender, prepare.highQC)
+                  isValid <- validateBlock(prepare.block)
```

Comment:
Does this realise the `safeNode` predicate? If so, it probably should depend on `highQC` as well. Just asking, it could be added in the relevant tasks

Reply:
No, that's not here, this is only validating whether the block contents are valid according to the application rules, so this is where we'd ask the PoW Interpreter whether any checkpoint candidates in the block are valid in the context of the state of the parent block (which has been validated before, evidenced by the High Q.C.).

The `safeNode` predicate is checked in [ProtocolState](https://github.com/input-output-hk/metronome/blob/develop/metronome/hotstuff/consensus/src/io/iohk/metronome/hotstuff/consensus/basic/ProtocolState.scala#L194) during the `Prepare` phase. That happens in `handleMessage` which already receives a validated block. 

There is some preliminary stateless validation in `ProtocolState.validateMessage`, but because that's stateless it doesn't check the extension rule (`lockedQC` depends on consensus state).

However, I added two more extension validations to the `ConsensusService`:
1. There's a [view number check](https://github.com/input-output-hk/metronome/blob/aa9e54de0a72a5db80650db1b88628abd78dd9fd/metronome/hotstuff/service/src/io/iohk/metronome/hotstuff/service/ConsensusService.scala#L88-L97) when the message is received to see that the leader is not proposing to use a quorum certificate that is older than our committed one.
2. There's a [parent check](https://github.com/input-output-hk/metronome/blob/aa9e54de0a72a5db80650db1b88628abd78dd9fd/metronome/hotstuff/service/src/io/iohk/metronome/hotstuff/service/ConsensusService.scala#L128-L131) before we hand over to the `SyncService` to see that the proposed block actually extends the High Q.C.

Based on these we can tell whether the block proposed to us looks valid extension of its parent and the previously committed state. Then the `SyncService` gets any missing dependencies and validates the content, replying with the outcome. Only then is the block delivered to the `ProtocolState` for consensus decision. By that time we may have moved on to a new view, but if everything checks out and it's a safe extension, a vote will be cast to it. 

This passing around the message realises this passage from the paper:

> During the protocol, a replica delivers a message only after the branch led by the node is already in its local tree.  In practice, a recipient who falls behind can catch up by fetching missing nodes from other replicas. 

In other words, by the time the consensus model sees the block, they should have already been vetted for their content. I'm just doing those extra 2 checks in the `ConsensusService` to throw out spam.

Reply:
This method has been [farmed out](https://github.com/input-output-hk/metronome/blob/aa9e54de0a72a5db80650db1b88628abd78dd9fd/metronome/hotstuff/service/src/io/iohk/metronome/hotstuff/service/SyncService.scala#L244) to the `ApplicationService` since this PR, so it's a bit clearer what the scope is.

Reply:
Thanks for clarifying